docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+What messages should be sent and timeouts triggered is defined mostly by the `Controller`. The exception is peer heartbeat mechanism which is `Executor`
+responsibility. The heartbeat mechanism consists of periodically sending `StatusRequest` messages to peers and signalling to `Executor` in case a peer does not respond 
+timely with the corresponding `StatusReport` message that a peer is not responsive so it should be removed from the peer list.    
+This simpler architecture is easier to understand and makes writing of unit tests simpler tasks as most of the critical
+logic is part of the `Controller` function. 
+
+### Implementation changes
+
+We assume the following system model for "fast sync" protocol: 
+
+* a node is connected to a random subset of all nodes that represents its peer set. Some nodes are correct and some might be faulty. We don't make assumptions about ratio of  
+  faulty nodes, i.e., it is possible that all nodes in some peer set are faulty. 
+* we assume that communication between correct nodes is synchronous, i.e., if a correct node `p` sends a message `m` to a correct node `q` at time `t`, then `q` will receive
+  message the latest at time `t+Delta` where `Delta` is a system parameter that is known. `Delta` is normally order of magnitude higher than the real communication delay 
+  (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `q` at time `t` and there is no the corresponding reply at time
+  `t + 2*Delta`, then `p` can assume that `q` is faulty. Note that the network assumptions for the consensus reactor are different 
+  (we assume partially synchronous model there). 
+
+The "fast sync" protocol is formally specified as follows:
+
+- `Correctness`: If a correct node `p` is connected to a correct node `q` for a long enough period of time, then `p` will eventually download all requested blocks from `q`.
+- `Termination`: If a set of peers of a correct node `p` is stable (no new nodes are added to the peer set) for a long enough period of time, then protocol eventually terminates. 
+- `Fairness`: A correct node `p` sends requests for blocks to all peers from its peer set.   
+
+The `Controller` can be modelled as a function with clearly defined inputs:
+
+* `State` - current state of the node. Contains data about connected peers and its behavior, pending requests, received blocks, etc.
+* `Event` - significant events in the network.
+
+producing clear outputs:
+
+* `State` - updated state of the node,
+* `Message` - signal what message to send,
+* `TimeoutTrigger` - signal that timeout should be triggered.
+
+```go
+type Event int
+
+const (
+	EventUnknown Event = iota
+	EventStatusReport
+	EventBlockRequest
+	EventNewBlock
+	EventRemovePeer
+	EventResponseTimeout
+	EventTerminationTimeout
+)
+
+type BlockRequest struct {
+	Height int64
+	PeerID p2p.ID
+}
+
+type NewBlock struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
```

Comment:
I think we need an overview of how all these data structures are used and when. Some control flow diagrams would be useful. 
Also, a part of this ADR should be the concurrency model/ design for scale/performance and testability. I believe this would be part of what you call the Executor task. I did some experiments with a version of the reactor poolRoutine that incorporates some of the Controller/ FSM changes. Sending one request per peer is slowing things down dramatically. Having hardcoded fixed limits may not be the best approach, maybe adaptive would work better. But in order to know, we need to have a good design (e.g. collect good stats, be able to abstract the things we sync, etc.) to understand what happens. Eyeing the logs is a nightmare. Not sure what we have today to collect operational data from nodes.

Reply:
I will try to add control diagram similar as we have for the current version. Maybe you can share source of the current diagram? Regarding performance concerns, extending this proposal to send multiple requests to a peer should be relatively straightforward. I will try to do it. I agree that number of pending requests per peer should be adaptive in ideal case but we need to estimate the complexity of such approach. We can maybe start with adding multiple pending requests (hardcoded value) to this proposal as this will let faster peer send multiple responses, and have them faster in position to serve new requests.